array: avoid overallocation when inserting at beginning of empty arrays

vtjnash opened the pull request:
array: reduce overallocation when inserting into small arrays

Fixes a minor nuisance where length 0 would grow to 8 and length 7 would also grow to 8. Instead growing by a flat 3x even initially.

Fixes #58640

vtjnash committed:
array: reduce overallocation when inserting into small arrays

Fixes a minor nuisance where length 0 would grow to 8 and length 7 would
also grow to 8. Instead growing by a flat 3x even initially.

Fixes #58640

oscardssmith commented on base/array.jl:
```diff
@@ -1110,22 +1110,23 @@ function _growbeg_internal!(a::Vector, delta::Int, len::Int)
-    if newoffset + newlen < memlen
+    # Check if we can reuse existing memory without allocating
+    # This is especially important for empty arrays where overallocation(0) might be larger than memlen
+    newmemlen = len + 2 * delta + 1
+    if newmemlen <= memlen
```
This isn't quite strict enough. We want to ensure that we only move the start of the array if there remains proportional space at both ends of the array (since otherwise you are signing up for potential quadratic behavior)

vtjnash replied:
Ah, you're right, I thought this seemed too easy. I changed it instead to use a different `overallocation` function logic. The current design does have a bit more "sawtooth" effect to it than we perhaps intended, so power of two sizes might grow much faster than the previous value (though in aggregate they grow over time equivalently):
```
r = 0:1024:1024^2; overallocation.(r) ./ r
```

oscardssmith replied:
I wonder if we should have a `-1` somewhere in the formula. Making power of 2 be the breaking point might be pushing us away from optimal sizes (e.g. multiples of a page).

vtjnash commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

oscardssmith commented:
Nanosoldier results seem noisy as expected.

vtjnash committed:
iobuffer: fix overflow OOB read/write bugs

maxsize is usually typemax, so need to be careful to not do comparisons after adding to it. Substracting from it should normally be perfectly fine. At worst we should compute a negative amount of space remaining.

vtjnash committed:
Update stream.jl
